Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift

```diff
@@ -0,0 +1,191 @@
+/// incur an allocation for every byte.
+///
+/// This delegate vends the response body as a `HTTBody` with one chunk for each
+/// `urlSession(_:didReceive data:)` callback. It also provides backpressure, which will
+/// suspend and resume the URLSession task based on a configurable high and low watermark.
+///
+/// When performing requests without a body, this delegate should be used with a
+/// `URLSessionDataTask` to stream the response body.
+///
+/// When performing requests with a body, this delegate should be used with a
+/// `URLSessionUploadTask` using `uploadTask(withStreamedRequest:delegate:)`, which will
+/// ask the delegate for a `InputStream` for the request body via the
+/// `urlSession(_:needNewBodyStreamForTask:)` callback.
+///
+/// The `urlSession(_:needNewBodyStreamForTask:)` callback will create a pair of bound
+/// streams, bridge the `HTTPBody` request body to the `OutputStream` and return the
+/// `InputStream` to URLSession. Backpressure for the request body stream is provided
+/// as an implementation detail of how URLSession reads from the `InputStream`.
+///
+/// Note that `urlSession(_:needNewBodyStreamForTask:)` may be called more than once, e.g.
+/// when performing a HTTP redirect, upon which the delegate is expected to create a new
+/// `InputStream` for the request body. This is only possible if the underlying `HTTPBody`
+/// request body can be iterated multiple times, i.e. `iterationBehavior == .multiple`.
+/// If the request body cannot be iterated multiple times, then the URLSession task will be cancelled.
+final class BidirectionalStreamingURLSessionDelegate: NSObject, URLSessionTaskDelegate, URLSessionDataDelegate {
+
+    let requestBody: HTTPBody?
+    var hasAlreadyIteratedRequestBody: Bool
+    var hasSuspendedURLSessionTask: Bool
+    let requestStreamBufferSize: Int
+    var requestStream: HTTPBodyOutputStreamBridge?
+
+    typealias ResponseContinuation = CheckedContinuation<URLResponse, any Error>
+    var responseContinuation: ResponseContinuation?
+
+    typealias ResponseBodyStream = AsyncBackpressuredStream<HTTPBody.ByteChunk, any Error>
+    var responseBodyStream: ResponseBodyStream
+    var responseBodyStreamSource: ResponseBodyStream.Source
+
+    /// This lock is taken for the duration of all delegate callbacks to protect the mutable delegate state.
+    ///
+    /// Although all the delegate callbacks are performed on the session's `delegateQueue`, there is no guarantee that
+    /// this is a _serial_ queue.
+    ///
+    /// Regardless of the type of delegate queue, URLSession will attempt to order the callbacks for each task in a
+    /// sensible way, but it cannot be guaranteed, specifically when the URLSession task is cancelled.
+    ///
+    /// Therefore, even though the `suspend()`, `resume()`, and `cancel()` URLSession methods are thread-safe, we need
+    /// to protect any mutable state within the delegate itself.
+    let callbackLock = NSLock()
```

Comment:
Do we prefer NSLock or NIOLock? I think both are used in some files

Reply:
This library doesn't depend on NIO.

Reply:
No, but it _has_ vendored in NIOLock for the AsyncBackpressuredStream. Let me see if I can switch to it.

Reply:
Switched to `NIOLock` in 84dd06e.